# TASK 4.9.1

## This script contains the following points:
### -  Importing libraries
### -  Importing Data
### -  Data Wrangling
### -  Data Consistency Checks
### -  Merge data 
### -  Export data

## 01. Importing libraries

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

## 02. Importing Data

In [2]:
path = r'C:\Users\Sanja\Documents\08-2020 Instacart Basket Analysis'

In [3]:
df_customer = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [4]:
ords_prods_agg = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data','orders_products_aggregated.pkl'))

## 03. Data Wrangling

In [5]:
df_customer.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [6]:
df_customer.shape

(206209, 10)

### Observation: 

#### - First Name can be renamed to 'first_name'
#### - Surnam can be renamed to 'last_name'
#### - Gender can be renamed to 'gender'
#### - STATE can be renamed to 'state'
#### - Age can be renamed to 'age'
#### - n_dependants can be renamed to 'number_of_dependents'
#### - fam_status can be renamed to 'marital_status'
#### - income can be renamed to 'spending_power'

### 3.1 Dropping Columns

### I've decided not to drop any columns.

### 3.2 Renaming Columns

In [7]:
df_customer.rename(columns = {'First Name' : 'first_name', 'Surnam' : 'last_name', 'Gender' : 'gender', 'STATE' : 'state', 'Age' : 'age', 'n_dependants' : 'number_of_dependents', 'fam_status' : 'marital_status', 'income' : 'spending_power' }, inplace = True)

In [8]:
df_customer.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependents marital_status  spending_power  
0                     3        married          165665  
1                     0         single           59285  
2                     2        married           99568  
3                     0         single           42049  
4                     1        married           40374

In [9]:
# Change of the data type of the columns from a number to a string to ignore when calculating descriptive statistics
df_customer['user_id'] = df_customer['user_id'].astype('str')

In [10]:
df_customer['user_id'].dtype

dtype('O')

## 04. Data Consistency Checks

### 4.1 Interpret the output of the df.describe() function

In [11]:
df_customer.describe()

age  number_of_dependents  spending_power
count  206209.000000         206209.000000   206209.000000
mean       49.501646              1.499823    94632.852548
std        18.480962              1.118433    42473.786988
min        18.000000              0.000000    25903.000000
25%        33.000000              0.000000    59874.000000
50%        49.000000              1.000000    93547.000000
75%        66.000000              3.000000   124244.000000
max        81.000000              3.000000   593901.000000

#### Investigating the frequency table, in particular the max and min section, i didn't find anyting strange about the data.
- age, result in a minimum of 18 and a maximum of 81
- number_of_dependents, result in a minimum of 0 and a maximum of 3.
- spending_power, result in a minimum of 25.903 and a maximum of 593.901.

### 4.2. Check for mixed-type data in your df_ords dataframe

In [12]:
# Check for mixed types
for col in df_customer.columns.tolist():
  weird = (df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customer[weird]) > 0:
    print (col)

first_name


#### Observation: The column 'first_name' contains mixed data type

In [13]:
df_customer['first_name'].dtypes

dtype('O')

### 4.3 Finding Missing Values

In [14]:
df_customer.isnull().sum()

user_id                     0
first_name              11259
last_name                   0
gender                      0
state                       0
age                         0
date_joined                 0
number_of_dependents        0
marital_status              0
spending_power              0
dtype: int64

In [15]:
df_nan = df_customer[df_customer['first_name'].isnull() == True]

In [16]:
df_nan

user_id first_name last_name  gender         state  age date_joined  \
53       76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73       13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82       89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99       96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105      29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...        ...        ...       ...     ...           ...  ...         ...   
206038  121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044  200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090  167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162  187532        NaN     Floyd  Female    California   39    4/1/2020   
206171  116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        number_of_dependents marital_status  spending_power  
53                         2        married           41709  
73                         0         single           82518  
82                         3        married          117099  
99                         1        married          155673  
105                        3        married          151819  
...                      ...            ...             ...  
206038                     3        married           87783  
206044                     2        married          108488  
206090                     1        married           45275  
206162                     0         single           56325  
206171                     2        married           59222  

[11259 rows x 10 columns]

#### Observation: mixed data type in column ' 'first_name' is due to 11259 'NaN' values, but i wouldn't delete those rows because the rest of them are holding meaningful information

### 4.4. Addressing Missing Values

In [17]:
# Replace NaN value with 'Unknown'
df_customer['first_name'].fillna('Unknown', inplace=True)

In [18]:
# cross-check  if there's mixed value types
for col in df_customer.columns.tolist():
  weird = (df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customer[weird]) > 0:
    print (col)

### There are no mixed data type in the dataframe anymore.

In [19]:
# Delete columns 'first_name' and 'last_name', we don't need the names for the analysis
customer = df_customer.drop(columns = ['first_name', 'last_name'])

### 4.5 Addressing Duplicates

In [20]:
customer_dups = customer[df_customer.duplicated()]

In [21]:
customer_dups

Empty DataFrame
Columns: [user_id, gender, state, age, date_joined, number_of_dependents, marital_status, spending_power]
Index: []

### No duplicates found.

In [22]:
customer.shape

(206209, 8)

## 05. Combine the customer data with the rest of your prepared Instacart data

#### Hint: Make sure the key columns are the same data type!

In [23]:
ords_prods_agg['user_id'].dtype

dtype('int64')

In [24]:
# Change of the data type of the columns from a string to a number, so the key columns are the same data type
customer['user_id'] = customer['user_id'].astype('int64')

In [25]:
customer['user_id'].dtype

dtype('int64')

In [26]:
ords_prods_agg.shape

(32404859, 23)

In [27]:
ords_prods_agg.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  ...  _merge        price_range    busiest_days  \
0         Soda  ...    both  Mid-range product  Regularly busy   
1         Soda  ...    both  Mid-range product     Slowes days   
2         Soda  ...    both  Mid-range product     Slowes days   
3         Soda  ...    both  Mid-range product     Slowes days   
4         Soda  ...    both  Mid-range product     Slowes days   

  busiest_period_of_day max_order  loyalty_flag mean_prices  spending_flag  \
0        Average orders        10  New customer    6.367797    Low spender   
1        Average orders        10  New customer    6.367797    Low spender   
2           Most orders        10  New customer    6.367797    Low spender   
3        Average orders        10  New customer    6.367797    Low spender   
4           Most orders        10  New customer    6.367797    Low spender   

  median_days_since_last_order         frequency_flag  
0                         20.5  Non-frequent customer  
1                         20.5  Non-frequent customer  
2                         20.5  Non-frequent customer  
3                         20.5  Non-frequent customer  
4                         20.5  Non-frequent customer  

[5 rows x 23 columns]

In [28]:
# Dropping column _merge from df 'ords_prods_agg', since we will get new  “_merge” column from the new merge
df_ords_prods_agg = ords_prods_agg.drop(columns = ['_merge'])

In [29]:
df_ords_prods_agg.shape

(32404859, 22)

In [30]:
customer.shape

(206209, 8)

In [31]:
# Merge df_customer and df_ords_prior using user_id as a key 
ords_prods_customer_merged = ords_prods_agg.merge(customer, on = 'user_id')

In [32]:
ords_prods_customer_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  ...  spending_flag  median_days_since_last_order  \
0         Soda  ...    Low spender                          20.5   
1         Soda  ...    Low spender                          20.5   
2         Soda  ...    Low spender                          20.5   
3         Soda  ...    Low spender                          20.5   
4         Soda  ...    Low spender                          20.5   

          frequency_flag  gender    state age date_joined  \
0  Non-frequent customer  Female  Alabama  31   2/17/2019   
1  Non-frequent customer  Female  Alabama  31   2/17/2019   
2  Non-frequent customer  Female  Alabama  31   2/17/2019   
3  Non-frequent customer  Female  Alabama  31   2/17/2019   
4  Non-frequent customer  Female  Alabama  31   2/17/2019   

   number_of_dependents marital_status  spending_power  
0                     3        married           40423  
1                     3        married           40423  
2                     3        married           40423  
3                     3        married           40423  
4                     3        married           40423  

[5 rows x 30 columns]

In [33]:
ords_prods_customer_merged.shape

(32404859, 30)

In [34]:
# Use the indicator argument to check whether there was a full match between the two dataframes
ords_prods_customer_merged_indicator = ords_prods_agg.merge(customer, on = 'user_id', indicator='merge_indicator')

In [35]:
ords_prods_customer_merged_indicator['merge_indicator'].value_counts()

both          32404859
left_only            0
right_only           0
Name: merge_indicator, dtype: int64

In [36]:
# Merge df_ords and df_ords_prior using order_id as a key & the argument how = 'outer', to double-check the full match 
df_merged_large_test= ords_prods_agg.merge(customer, on = 'user_id', indicator = 'merge_indicator_1', how = 'outer')

In [37]:
# Use the indicator argument to check whether there was a full match between the two dataframes
df_merged_large_test['merge_indicator_1'].value_counts()

both          32404859
left_only            0
right_only           0
Name: merge_indicator_1, dtype: int64

#### Note: After using this method to double-check the merge, we can see that we actually have a full match. 

##  06. Exporting Data in Pickle Format

In [38]:
# Final check of the columns
ords_prods_customer_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 30 columns):
 #   Column                        Dtype   
---  ------                        -----   
 0   order_id                      int64   
 1   user_id                       int64   
 2   order_number                  int64   
 3   orders_day_of_week            int64   
 4   order_hour_of_day             int64   
 5   days_since_last_order         float64 
 6   product_id                    int64   
 7   add_to_cart_order             int64   
 8   reordered                     int64   
 9   product_name                  object  
 10  aisle_id                      int64   
 11  department_id                 int64   
 12  prices                        float64 
 13  _merge                        category
 14  price_range                   object  
 15  busiest_days                  object  
 16  busiest_period_of_day         object  
 17  max_order                     int64   
 18  

In [39]:
# Dropping column _merge from final dataset 'ords_prods_customer_merged'
orders_products_all = ords_prods_customer_merged.drop(columns = ['_merge'])

In [41]:
orders_products_all.shape

(32404859, 29)

In [40]:
orders_products_all.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_all.pkl'))